### ISSUE TITLE
Retrieve parent resolver context into sub-query resolver

### ISSUE THREAD

**Issue by Vichoko** [strawberry-graphql/strawberry#1153]
## TL;DR
**The use case I'm facing is that I want access to some data computed on the parent resolver in the child resolver.**

## Description
If we have a query like this:

```graphql
query {
  library (name: "Borders Library") {
    id
    books (genre: "horror") {
      name
      code
    }
}
```
The computation of this single query is resolved from an outside-inside fashion. Where the parent `library` object type is resolved first, then the `books` child object types are resolved at last.

In this child `books` resolver, for instance, we might want to sub-set our books based on the library name and book genre.
As the computation flows from outside to inside, it feels pretty natural that it might exist a way to record some notion of state/context in the parent resolver, in a way that it could be accessible from the child resolver to help with the computations.

As far as I know, there's no way to accomplish this in the current state of the framework. I know it would be fragile to implement some ad-hoc logic for this purpose as in a multi-user case it's hard to pass context from one block of code to another, where this will most likely fall into a multiple producer-consumer problem.

## Proposed solution
There's one place I think this would fit well, and it's on the `info: Info` object passed to every resolver.

There could be some dict-like or something where you could put stuff in the parent resolver, that might be accessible from the child resolver's `info` parameter.

### One resolver to rule them all
The most simple solution I can think of is to move all inputs into the parent query like this:
```graphql
query {
  library (name: "Borders Library", book_genre: "horror") {
    id
    books {
      name
      code
    }
}
```

And then I can use the parent resolver to compute and populate all fields in one single resolver. But, I feel this solution is no so elegant.

### Hacky resolving
By using the info object I can access the raw query, so I could implement the previous solution without moving the books inputs into the parent. In other words, I can retrieve those parameters from the parent resolver from the raw query, and solve everything in one resolver, but making the user think they are different inputs.

### Context object
I was thinking of implementing a singleton object to publish some data that the resolvers could access. But the more I think about it, the more complex it sounds because there might be multiple producers and multiple readers. I can't think of a simple way to coordinate this without unexpected edge cases.

## Questions

I want to know if there's any abstraction that might help me with this use case. 

Also as there are no abstractions yet, maybe my approach is a code-smell symptom so feel free to discuss with me any alternatives.

**Comment by jkimbo** [comment]
@Vichoko what might help you is `strawberry.Private` which lets you attach extra information to types without it being exposed in the schema (documentation is coming #450 ).

So your query might be implemented like this:

```python
import strawberry

@strawberry.type
class Library:
	id: str
	name: strawberry.Private[str]  # Don't expose this to the schema
	
	@strawberry.field
	def books(self, genre: str) -> List[Book]:
		books = get_books(
			library_name=self.name,  # use the library name here to get a list of books
			genre=genre,
		)
		return [Book(name=book.name, code=book.code) for book in books]

@strawberry.type
class Query:
	@strawberry.field
	def library(self, name: str) -> Library:
		return Library(id="", name=name)
```

By using `strawberry.Private` you can avoid having to store anything on the `context` object. As you rightly point out it could lead to some weird bugs.

Another thing to note: `strawberry.Private` can annotate any type so you could pass the whole DB instance to the `Library` type and make use of it in the `books` resolver. Thats what I do when integrating with Django models and it works well.

Hope that helps!

**Comment by Vichoko** [comment]
This is wonderful. It was exactly what I needed! 
Thank you so much. ^^

**Comment by jkimbo** [comment]
Glad I could help!

**Comment by Vichoko** [comment]
It works smoothly. But I wanted to note that I found a bug that causes the app to crash when using any field with strawberry.Private together with `from  __future__ import annotations`.

It throws the following stack trace:

`Error
Traceback (most recent call last):
  File "/Users/vicenteoyanedel/opt/miniconda3/envs/main/lib/python3.9/unittest/loader.py", line 34, in testFailure
    raise self._exception
ImportError: Failed to import test module: tests.unit.recommender.test_schema
Traceback (most recent call last):
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/graphql/type/definition.py", line 767, in fields
    fields = resolve_thunk(self._fields)
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/graphql/type/definition.py", line 296, in resolve_thunk
    return thunk() if callable(thunk) else thunk
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/strawberry/schema/schema_converter.py", line 277, in get_graphql_fields
    graphql_fields[field_name] = self.from_field(field)
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/strawberry/schema/schema_converter.py", line 128, in from_field
    if isinstance(field.type, StrawberryOptional):
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/strawberry/field.py", line 174, in type
    return self.type_annotation.resolve()
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/strawberry/annotation.py", line 60, in resolve
    evaled_type = _eval_type(annotation, self.namespace, None)
  File "/Users/vicenteoyanedel/opt/miniconda3/envs/main/lib/python3.9/typing.py", line 290, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "/Users/vicenteoyanedel/opt/miniconda3/envs/main/lib/python3.9/typing.py", line 545, in _evaluate
    type_ =_type_check(
  File "/Users/vicenteoyanedel/opt/miniconda3/envs/main/lib/python3.9/typing.py", line 164, in _type_check
    raise TypeError(f"{msg} Got {arg!r:.100}.")
TypeError: Forward references must evaluate to types. Got strawberry.Private[bool].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/vicenteoyanedel/opt/miniconda3/envs/main/lib/python3.9/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/Users/vicenteoyanedel/opt/miniconda3/envs/main/lib/python3.9/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/Users/vicenteoyanedel/git/recommender/tests/unit/recommender/test_schema.py", line 5, in <module>
    from recommender.schema import schema
  File "/Users/vicenteoyanedel/git/recommender/recommender/schema.py", line 24, in <module>
    schema = strawberry.Schema(query=Query)
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/strawberry/schema/schema.py", line 74, in __init__
    self._schema = GraphQLSchema(
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/graphql/type/schema.py", line 208, in __init__
    collect_referenced_types(query)
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/graphql/type/schema.py", line 423, in collect_referenced_types
    collect_referenced_types(field.type)
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/graphql/type/schema.py", line 422, in collect_referenced_types
    for field in named_type.fields.values():
  File "/Users/vicenteoyanedel/opt/miniconda3/envs/main/lib/python3.9/functools.py", line 969, in __get__
    val = self.func(instance)
  File "/Users/vicenteoyanedel/git/recommender/.venv/lib/python3.9/site-packages/graphql/type/definition.py", line 769, in fields
    raise TypeError(f"{self.name} fields cannot be resolved. {error}")
TypeError: SimilarFoodInstances fields cannot be resolved. Forward references must evaluate to types. Got strawberry.Private[bool].`

**Comment by jkimbo** [comment]
Ah yeah there is a bug open for that: https://github.com/strawberry-graphql/strawberry/issues/843

**Comment by axiomofjoy** [comment]
Is there a way of reading the subquery from a parent resolver? In my case, I want to make a single DB call in the parent resolver to grab all of the information and store off DB objects in a `Private` field on the subquery types to be returned by the subquery resolver. I don't want to over-fetch if a particular subquery is not present, though. In the example above, imagine the `library` resolver made a single DB call that will also fetch the information needed for the `books` resolver, if this subquery was present, but won't fetch this information otherwise.
